### ISSUE TITLE
Freesing/crash JVM in multithreading for a lot of UMat

### ISSUE THREAD

**Issue by Daniel-Alievsky** [bytedeco/javacpp-presets#663]
Can you check the following very simple test:

```
import org.bytedeco.javacpp.opencv_core;

public class MultitrehadingManyUMatProblem {
    public static void main(String[] args) {
        System.out.println("OpenCL existance: " + opencv_core.haveOpenCL());
        System.out.println("OpenCL usage: " + opencv_core.useOpenCL());
        System.out.println();
        final int n = 300;
        for (int k = 0; k < n; k++) {
            System.out.println("Starting " + k + "...");
            final int index = k;
            new Thread(() -> {
                final opencv_core.UMat mat = new opencv_core.UMat(2000, 2000, opencv_core.CV_8U,
                        new opencv_core.Scalar(0, 0, 0, 0));
                System.out.println("Preparing " + index + "...");
                final opencv_core.UMat clone = mat.clone();
                System.out.println("Finishing " + index + "...");
            }).start();
        }
    }
}

```
Please try to call this several time with changing number of iterations n (100, 1000, 2000)...
I saw following variants of behavior with different n.
Sometimes (for n=300 on my computer) it just freezes and stop to react after allocating some number of UMat. 
Sometimes I see different exception messages like CL_MEM_OBJECT_ALLOCATION_FAILURE and the program also freezes.
Sometimes I see the messages like 
[ERROR:270] Can't create file: c:\temp\opencv\3.4.0\opencl_cache\NVIDIA_Corporation--GeForce_GTX_770M--417_35NVIDIA_Corporation--GeForce_GTX_770M--417_35NVIDIA_Corporation--GeForce_GTX_770M--417_35NVIDIA_Corporation--GeForce_GTX_770M--417_35/core--copyset_f8f028f1776dc5c98bf03411d3b72318.bin
At least one time the was low-level JVM crash.

Is it possible to fix this anyhow? I understand that allocating too large GPU memory _should_ lead to exceptions like "OutOfMemory". But locking and freezing all execution is obviously abnormal behavior. Imagine that it will occur on some high-performance server, that uses GPU for solving "heavy" customer tasks. Low-level JVM crash is also incorrect behavior.

The most serious problem is freezing; in this case, we cannot detect a problem in any way and cannot inform the user that there are insufficient resources. Probably, it is some OpenCV or JavaCPP synchronization bug?

**Comment by saudet** [comment]
To fix this, you'll need to call the `deallocate()` method either manually, with try-with-resources statements, or via `PointerScope`: http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/

If the application locks up though, that's an issue with your OpenCL implementation, same as issue #658.

**Comment by Daniel-Alievsky** [comment]
Maybe you didn't understand me. Of course, I need to deallocate all used memory. But I tested situation when the application really needs a lot of memory. For example, it is an application server, solving complex image processing tasks (our servers process panoramas ~100000x100000 pixels, splitting them to frames 2000x2000 or 8000x8000). GPU optimization is very important in such solutions. In a single-user mode, usually there are no problems, but in multiuser mode it is very possible that the system will be sometimes overloaded.

My question is about such situation, when the amount of GPU memory is insufficient. In a case of usual RAM, this problem is well-known and has stable solutions. In particular, JVM and many other subsystems have a total allowed memory limit and report about out of memory and reject executing the corresponding task. Other applications will use more and more memory, and OS will start disk swapping: the server will slow down, but continue work. In any case, the user or server admin can detect the problem and make necessary efforts.

But current implementation of JavaCPP/OpenCV provide no chances to do something reasonable. The system will just freeze! No any signals about problem, no any log messages, no error codes informing about dangerous situation. 

I believe here is just a bug like deadlock, when several threads try to allocate insufficient memory. And it should be fixed. Please replace `.start()` with `.run()` in my test and add little delay:

```
import org.bytedeco.javacpp.opencv_core;

public class MultitrehadingManyUMatProblem {
    public static void main(String[] args) {
        System.out.println("OpenCL existance: " + opencv_core.haveOpenCL());
        System.out.println("OpenCL usage: " + opencv_core.useOpenCL());
        System.out.println();
        final int n = 300;
        for (int k = 0; k < n; k++) {
            System.out.println("Starting " + k + "...");
            final int index = k;
            new Thread(() -> {
                final opencv_core.UMat mat = new opencv_core.UMat(2000, 2000, opencv_core.CV_8U,
                        new opencv_core.Scalar(0, 0, 0, 0));
                System.out.println("Preparing " + index + "...");
                final opencv_core.UMat clone = mat.clone();
                System.out.println("Finishing " + index + "...");
                try {
                    Thread.sleep(200);
                } catch (InterruptedException e) {
                }
            }).run();
        }
    }
}
```
Now we have single-thread version: we just allocate 2*300 UMats 2000x2000 (`mat` and `clone`), i.e. ~2.4 GB GPU memory. Really, Process Explorer shows that the process allocated ~2.4 GB GPU memory. My computer has 3 GB GPU memory, and all works fine! (If your computer has other amount, just correct "n" constant to allocate 70-80% of your GPU memory.)

Then replace "run()" back with "start()". On my computer, all JVM freezes after printing
....
Finishing 239...
Finishing 33...
Preparing 296...
Preparing 264...
Process Explorer shows that the process allocated only 1.2 GB GPU. And this bug is absolutely stable: the system freezes every time, though we used less than 50% GPU memory.

And sometimes all JVM crashes at the low level. Below is the beginning of the dump-file hs_err_pid5952.log, written by JVM after such crash:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffa627a0d01, pid=5952, tid=0x0000000000001e1c
#
# JRE version: Java(TM) SE Runtime Environment (8.0_181-b13) (build 1.8.0_181-b13)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.181-b13 mixed mode windows-amd64 compressed oops)
# Problematic frame:
# C  [opencv_core340.dll+0x150d01]
#
# Failed to write core dump. Minidumps are not enabled by default on client versions of Windows
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  T H R E A D  ---------------

Current thread (0x000000001ec6c800):  JavaThread "Thread-73" [_thread_in_native, id=7708, stack(0x0000000026640000,0x0000000026740000)]

siginfo: ExceptionCode=0xc0000005, reading address 0xffffffffffffffff

Registers:
RAX=0x005f00001fdfc880, RBX=0x000000002673e8f8, RCX=0x00000000000000bf, RDX=0x00043ffffff01ff9
RSP=0x000000002673e870, RBP=0x000000002673e8f9, RSI=0x0000000000000007, RDI=0x000000002673e908
R8 =0x00000000003f1380, R9 =0x0000000000008000, R10=0x000000001fdb8f30, R11=0x000000001fdb8fb0
R12=0x00000000ffffffff, R13=0x0000000000000004, R14=0x00000000002dbd00, R15=0x0000000020df0b90
RIP=0x00007ffa627a0d01, EFLAGS=0x0000000000010202

Top of Stack: (sp=0x000000002673e870)
0x000000002673e870:   000000002673e908 0000000000000000
0x000000002673e880:   000000002673e8f9 0000000000000000
0x000000002673e890:   000000002673e880 0000000400000001
0x000000002673e8a0:   0000000000000000 0000000000000000
0x000000002673e8b0:   fffffffffffffffe 0000000000000000
0x000000002673e8c0:   0000000000000000 0000000000000000
0x000000002673e8d0:   0000000000000000 0000000000000000
0x000000002673e8e0:   0000000000000000 0000000000000000
0x000000002673e8f0:   0000000000000000 0000000000000000
0x000000002673e900:   0000000000000000 0000000000000000
0x000000002673e910:   0000000000000000 000000001fdb8f00
0x000000002673e920:   000000001b54ac00 0000000000000000
0x000000002673e930:   000000000000000f 00009ddd196fc393
0x000000002673e940:   00000000002dbd00 0000000020df0b00
0x000000002673e950:   000000002673ea60 00007ffa627a12f2
0x000000002673e960:   0000000020df0bd0 00000000215c1444 

Instructions: (pc=0x00007ffa627a0d01)
0x00007ffa627a0ce1:   f9 1f f0 ff ff 3f 04 00 0f 1f 80 00 00 00 00 49
0x00007ffa627a0cf1:   83 7e 40 10 72 06 49 8b 46 28 eb 04 49 8d 46 28
0x00007ffa627a0d01:   0f b6 0c 30 8d 41 d3 3c 32 77 06 48 0f a3 c2 72
0x00007ffa627a0d11:   1d 80 e9 61 80 f9 19 76 15 49 83 7e 40 10 72 06 


Register to memory mapping:

RAX=0x005f00001fdfc880 is an unknown value
RBX=0x000000002673e8f8 is pointing into the stack for thread: 0x000000001ec6c800
RCX=0x00000000000000bf is an unknown value
RDX=0x00043ffffff01ff9 is an unknown value
RSP=0x000000002673e870 is pointing into the stack for thread: 0x000000001ec6c800
RBP=0x000000002673e8f9 is pointing into the stack for thread: 0x000000001ec6c800
RSI=0x0000000000000007 is an unknown value
RDI=0x000000002673e908 is pointing into the stack for thread: 0x000000001ec6c800
R8 =0x00000000003f1380 is an unknown value
R9 =0x0000000000008000 is an unknown value
R10=0x000000001fdb8f30 is an unknown value
R11=0x000000001fdb8fb0 is an unknown value
R12=0x00000000ffffffff is an unknown value
R13=0x0000000000000004 is an unknown value
R14=0x00000000002dbd00 is an unknown value
R15=0x0000000020df0b90 is an unknown value


Stack: [0x0000000026640000,0x0000000026740000],  sp=0x000000002673e870,  free space=1018k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [opencv_core340.dll+0x150d01]
C  [opencv_core340.dll+0x1512f2]
C  [opencv_core340.dll+0x151120]
C  [opencv_core340.dll+0x14b251]
C  [opencv_core340.dll+0x13fe1f]
C  [opencv_core340.dll+0x1c5cb2]
C  [opencv_core340.dll+0x1c0efb]
C  [jniopencv_core.dll+0x2d31]
C  [jniopencv_core.dll+0x30a29]
C  0x0000000001ff8c67

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  org.bytedeco.javacpp.opencv_core$UMat.allocate(IIILorg/bytedeco/javacpp/opencv_core$Scalar;)V+0
j  org.bytedeco.javacpp.opencv_core$UMat.<init>(IIILorg/bytedeco/javacpp/opencv_core$Scalar;)V+14
j  com.siams.stare.extensions.javacpp.opencv.MultitrehadingManyUMatProblem.lambda$main$0(I)V+22
j  com.siams.stare.extensions.javacpp.opencv.MultitrehadingManyUMatProblem$$Lambda$1.run()V+4
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub


```

**Comment by mnogono** [comment]
[ WARN:2] Detected OpenCL cache directory for other version of OpenCL device. We assume that these directories are obsolete after OpenCL runtime/drivers upgrade.
[ WARN:2] Trying to remove these directories...
[ WARN:2] - NVIDIA_Corporation--GeForce_GTX_760--417_22NVIDIA_Corporation--GeForce_GTX_760--417_22
[ WARN:2] Note: You can disable this behavior via this option: OPENCV_OPENCL_CACHE_CLEANUP=0
[ INFO:0] Initialize OpenCL runtime...
[ INFO:1] Successfully initialized OpenCL cache directory: C:\Users\semen\AppData\Local\Temp\1\opencv\3.4.0\opencl_cache\
[ INFO:2] Preparing OpenCL cache configuration for context: NVIDIA_Corporation--GeForce_GTX_760--417_22NVIDIA_Corporation--GeForce_GTX_760--417_22NVIDIA_Corporation--GeForce_GTX_760--417_22
[ WARN:2] Removed: C:\Users\semen\AppData\Local\Temp\1\opencv\3.4.0\opencl_cache\NVIDIA_Corporation--GeForce_GTX_760--417_22NVIDIA_Corporation--GeForce_GTX_760--417_22

**Comment by Daniel-Alievsky** [comment]
I can add that JVM freezes on my computer also for n=50: 400 MB from available 3 GB, less than 15%.

It is obviously a bug: we _should_ be able to use 15% of available GPU in parallel threads without any problems. 

Interesting that on my computer the problem disappeared when I comment the line 
  `final opencv_core.UMat clone = mat.clone();`
Without cloning, the test successfully allocates requested number of UMats! (Or shows exceptions, that there is not enough memory, if n is too high.) I tested ~10 times, and didn't see freezing.

It seems, there is some deadlock, maybe in cloning GPU or some other operations. Maybe you just need to globally synchronize something. I think it makes sense to reopen this issue.

**Comment by saudet** [comment]
I still don't think it's related to JavaCPP. Try at least to use a
different OpenCL implementation.

**Comment by Daniel-Alievsky** [comment]
We already tried another implementation: see comment of  mnogono 

But when I testthe same code in last SNAPSHOT version (JavaCPP 1.4.4 and OpenCVV 4.0.0), I cannot reproduce the freezing. Maybe authors of OpenCV fixed this problem?

In JavaCPP 1.4 + OpenCV 3.4.0, the freezing is a very stable bug, occurring while usage 10-15% of available GPU memory. In this state, Java thread dump shows that all threads are "working" inside native method "UMat.clone", I tried to replace simple `.clone()` with the next method
```
    public static opencv_core.UMat clone(opencv_core.UMat u) {
        opencv_core.UMat result = new opencv_core.UMat(u.rows(), u.cols(), u.type());
        u.copyTo(result);
        return result;
    }

```
It does not help; now, all threads freeze in "copyTo" method. The correct workaround is the following:
```
    public static opencv_core.UMat clone(opencv_core.UMat u) {
        try (opencv_core.Mat cloneInRAM = new opencv_core.Mat(u.rows(), u.cols(), u.type())) {
            u.copyTo(cloneInRAM);
            opencv_core.UMat result = new opencv_core.UMat(u.rows(), u.cols(), u.type());
            u.copyTo(result);
            return result;
        }
    }

```
But I hope that this problem was really fixed in OpenCV 4.0.0, and we can forget about this. How do you think?

**Comment by saudet** [comment]
Yes, OpenCV 3.4.0 is pretty old, a lot has been fixed since.
